tsl/src/nodes/decompress_chunk/decompress_chunk.c

```diff
@@ -560,6 +560,89 @@ can_sorted_merge_append(PlannerInfo *root, CompressionInfo *info, Chunk *chunk)
 	return merge_result;
 }
 
+/*
+ * If at least one chunk of a hypertable is uncompressed, PostgreSQL will generate a MergeAppend
+ * path in generate_orderedappend_paths() / create_merge_append_path() due to the existing pathkeys
+ * of the index on the uncompressed chunk. If no path keys are present, no MergeAppend path is
+ * generated by PostgreSQL. In that case, the ChunkAppend optimization cannot be used because
+ * MergeAppend path can be promoted in ts_chunk_append_path_create().
+ *
+ * Especially limit queries that use an ORDER BY that is compatible with the partitioning of the
+ * hypertable could be inefficiently executed so far. In that case, an expensive query plan with a
+ * sort node on top of the append node could be chosen. Due to the sort node at the high level in
+ * the query plan and the missing ChunkAppend node, all chunks are decompressed (instead of only the
+ * actually needed ones).
+ *
+ * This function adds a sorted path to the chunk if this is beneficial. As a result, this chunk has
+ * path keys and PostgreSQL will generate a MergeAppendPath that can be performed by ChunkAppend.
+ *
+ * The logic is inspired by PostgreSQL's add_paths_with_pathkeys_for_rel() function.
+ */
+static void
+add_chunk_sorted_path(PlannerInfo *root, RelOptInfo *chunk_rel, Hypertable *ht, Chunk *chunk,
+					  Index ht_relid, Path *decompress_chunk_path)
+{
+	if (root->parse->sortClause == NIL || root->cte_plan_ids != NIL ||
```

Comment:
`root->cte_plan_ids != NIL` is added to prevent a crash in the IndexScan node under DecompressChunk when CTEs are used (`econtext->ecxt_scantuple` is `NULL` in some cases). This crash is currently under investigation.